rpc/src/v1/helpers/signing_queue.rs

```diff
@@ -100,86 +99,77 @@ pub trait SigningQueue: Send + Sync {
 
 #[derive(Debug, Clone, PartialEq)]
 /// Result of a pending confirmation request.
-pub enum ConfirmationResult {
-	/// The request has not yet been confirmed nor rejected.
-	Waiting,
+// TODO [kirushik] Let's get rid of this completely, and simplify that `Result<Option<Result...` juggling somehow
+pub enum ConfirmationOutcome {
 	/// The request has been rejected.
 	Rejected,
 	/// The request has been confirmed.
-	Confirmed(RpcResult),
+	Confirmed(ConfirmationResponse),
+	/// We've failed to determine the state of the confirmation
+	Failed(Error),
+}
+
+impl From<Option<RpcResult>> for ConfirmationOutcome {
+	fn from(result: Option<RpcResult>) -> Self {
+		match result {
+			None => ConfirmationOutcome::Rejected,
+			Some(Ok(response)) => ConfirmationOutcome::Confirmed(response),
+			Some(Err(error)) => ConfirmationOutcome::Failed(error),
+		}
+	}
 }
 
-type Listener = Box<FnMut(Option<RpcResult>) + Send>;
+impl Into<Result<Option<ConfirmationResponse>, Error>> for ConfirmationOutcome {
+	fn into(self) -> Result<Option<ConfirmationResponse>, Error> {
+		match self {
+			ConfirmationOutcome::Rejected => Err(errors::request_rejected()),
+			ConfirmationOutcome::Confirmed(confirmation) => Ok(Some(confirmation)),
+			ConfirmationOutcome::Failed(error) => Err(error),
+		}
+	}
+}
 
-/// A handle to submitted request.
-/// Allows to block and wait for a resolution of that request.
-pub struct ConfirmationToken {
-	result: Arc<Mutex<ConfirmationResult>>,
-	listeners: Arc<Mutex<Vec<Listener>>>,
+struct ConfirmationSender {
+	sender: oneshot::Sender<ConfirmationOutcome>,
 	request: ConfirmationRequest,
 }
 
-pub struct ConfirmationPromise {
+/// Receiving end of the Confirmation channel; can be used as a `Future` to await for `ConfirmationRequest`
+/// being processed and turned into `ConfirmationOutcome`
+#[must_use = "futures do nothing unless polled"]
+pub struct ConfirmationReceiver {
```

Comment:
Do we need that structure at all? Maybe just return `(U256, Receiver)` instead of wrapping it.

Reply:
I'll try and see if it will lead to the more elegant code — I'm using `id()` method here and there, and I would need to replace it with explicit id passing.